Source/SwiftLintFramework/Rules/Idiomatic/NoMagicNumbersRule.swift

```diff
@@ -0,0 +1,85 @@
+import SwiftSyntax
+
+public struct NoMagicNumbersRule: SwiftSyntaxRule, OptInRule, ConfigurationProviderRule {
+    public init() {}
+
+    public var configuration = SeverityConfiguration(.warning)
+
+    public static let description = RuleDescription(
+        identifier: "no_magic_numbers",
+        name: "No Magic Numbers",
+        description: "Magic numbers should be replaced by named constants.",
+        kind: .idiomatic,
+        nonTriggeringExamples: [
+            Example("var foo = 123"),
+            Example("static let bar: Double = 0.123"),
+            Example("array[0] + array[1] "),
+            Example("let foo = 1_000.000_01"),
+            Example("// array[1337]"),
+            Example("baz(\"9999\")"),
+            Example("""
+        func foo() {
+            let x: Int = 2
+            let y = 3
+            let vector = [x, y, -1]
+        }
+        """),
+            Example("""
+        class A {
+            var foo: Double = 132
+            static let bar: Double = 0.98
+        }
+        """),
+            Example("""
+        @available(iOS 13, *)
+        func version() {
+            if #available(iOS 13, OSX 10.10, *) {
+                return
+            }
+        }
+        """)
+        ],
+        triggeringExamples: [
+            Example("foo(↓321)"),
+            Example("bar(↓1_000.005_01)"),
+            Example("array[↓42]"),
+            Example("let box = array[↓12 + ↓14]"),
+            Example("let a = b + 2.0"),
+            Example("Color.primary.opacity(isAnimate ? ↓0.1 : ↓1.5)")
+        ]
+    )
+
+    public func makeVisitor(file: SwiftLintFile) -> ViolationsSyntaxVisitor {
+        Visitor(viewMode: .sourceAccurate)
+    }
+}
+
+private extension NoMagicNumbersRule {
+    final class Visitor: ViolationsSyntaxVisitor {
+        override func visitPost(_ node: FloatLiteralExprSyntax) {
+            if node.floatingDigits.isMagicNumber {
+                self.violations.append(node.floatingDigits.positionAfterSkippingLeadingTrivia)
+            }
+        }
+
+        override func visitPost(_ node: IntegerLiteralExprSyntax) {
+            if node.digits.isMagicNumber {
+                self.violations.append(node.digits.positionAfterSkippingLeadingTrivia)
+            }
+        }
+    }
+}
+
+private extension TokenSyntax {
+    var isMagicNumber: Bool {
+        let numerStr = text.replacingOccurrences(of: "_", with: "")
+
+        guard let number = Double(numerStr),
+              ![0, 1].contains(number),
+              let parentToken = parent?.parent,
+              !parentToken.is(InitializerClauseSyntax.self) else {
```

Comment:
Are you still thinking about excluding annotations like `@available` here too, and, as you suggested, things like `let i = Int64(10)`?

Reply:
`@available` is apparently being ignored (see also test case).
With the other one I'm a bit unsure bc of how specific it is.. also, for cases like that `let i: Int64 = 10` could theoretically be used to avoid the warning. wdyt?

Reply:
I see. The numbers in the `@available` attributes are special syntax elements. That's why they are automatically excluded.

Regarding the second part, you are right in that it's not obvious which case must be allowed and which should be ignored. Let's keep it how it is for now and come back to it when there are specific requirements.